### PR TITLE
[CDAP-6745] [CDAP-6174] Integrates log viewer with tracker

### PR DESCRIPTION
**Pull request by PGuimarais**
tracker full screen view raw
- Integrates LogViewer with tracker
- Provides full-screen solution for navigator
- Changes artifact version
- 'View Raw' now renders Raw Log data in modal window

Fixes Bugs:
- https://issues.cask.co/browse/CDAP-6745
- https://issues.cask.co/browse/CDAP-5339
- https://issues.cask.co/browse/CDAP-6840
- https://issues.cask.co/browse/CDAP-6713
- https://issues.cask.co/browse/CDAP-6760

Bamboo Build:
http://builds.cask.co/browse/CDAP-DRC4151

### CODE REVIEW

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 1]
```diff
@@ -33,6 +34,7 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
     this.totalCount = 0;
     this.fullScreen = false;
     this.applicationIsRunning = false;
+    this.programStatus = 'STOPPED';
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: We should probably no default this to STOPPED. The default is not STOPPED.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 2]
```diff
@@ -516,7 +546,9 @@ angular.module(PKG.name + '.commons')
         programType: '@',
         programId: '@',
         runId: '@',
-        getDownloadFilename: '&'
+        getDownloadFilename: '&',
+        integratedWith: '@',
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Is this attribute being used anywhere?

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 3]
```diff
@@ -516,7 +546,9 @@ angular.module(PKG.name + '.commons')
...
+        statusName: '@'
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: This is actually the entity or log name right? ProgramId or app ID or `Cloudera Navigator`.

**Comment by ajainarayanan** [comment]
@PGuimarais Seeing this in workflows,
screen shot 2016-08-01 at 11 35 28 am

I think we should probably default to the entity ID or just don't show the status if there is no title. I thin the former would be more useful. Just seeing the status alone kind of looks like its broken.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 4]
```diff
@@ -380,8 +398,29 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
       });
   };
 
-  function formatDate(date, isDownload) {
+  const setProgramStatus = (status) => {
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: This could be `setProgramMetadata` and that can set both name and status.

File: cdap-ui/app/directives/log-viewer/log-viewer.js (hunk 1 above)
Comment (PGuimarais) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Now defaults to 'Not Started'

File: cdap-ui/app/directives/log-viewer/log-viewer.js (hunk 2 above)
Comment (PGuimarais) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: No, removed

File: cdap-ui/app/directives/log-viewer/log-viewer.js (hunk 3 above)
Comment (PGuimarais) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Correct

File: cdap-ui/app/directives/log-viewer/log-viewer.js (hunk 4 above)
Comment (PGuimarais) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Fixed. Now sets entity name.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 5]
```diff
@@ -14,17 +14,19 @@
...
   var dataSrc = new MyCDAPDataSource($scope);
   var pollPromise;
   //Collapsing LogViewer Table Columns
   var columnsList = [];
   var collapseCount = 0;
+  this.$uibModal = $uibModal;
 
   this.setDefault = () => {
     this.textFile = null;
+    this.statusType = 0;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: if you are setting the program status default as 'Not Started', this statusType should be 3. Otherwise it will say status Not Started but the circle is green

File: cdap-ui/app/directives/log-viewer/log-viewer.less [hunk 6]
```diff
@@ -32,6 +32,80 @@
...
+    padding: 0 20px;
+  }
+
+  .modal-footer {
+    padding-left: 20px;
+    padding-right: 20px;
+  }
+}
+
+.modal-dialog {
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.less]: why 2 .modal-dialog?

**Comment by ajainarayanan** [comment]
Talked to @PGuimarais 
- All program types, if no runs, should show the program name & status as 'Not Started' 
- View Raw button should not make a backend call when no runs
- Download button should be disabled on no runs.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 7]
```diff
@@ -14,17 +14,49 @@
...
   var pollPromise;
   //Collapsing LogViewer Table Columns
   var columnsList = [];
   var collapseCount = 0;
+  this.$uibModal = $uibModal;
+
+  this.setProgramMetadata = (status) => {
+    this.programStatus = status;
+
+    if(this.entityName && this.entityName.length === 0) {
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: checking this.entityName should be sufficient, right?

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 8]
```diff
@@ -64,6 +97,63 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
+      this.toggleMaximizedView = (isExpanded) => {
+        this.windowMode = (isExpanded) ? 'expand' : 'regular';
+      };
+
+      this.windowMode = 'regular';
+
+      modalDataSrc.request({
+        _cdapNsPath: `/apps/${rAppId}/${rProgramType}/${rProgramId}/runs/${rRunId}/logs?start=${rStartTimeSec}`
+      }).then((res) => {
+        if(res === undefined || res.length === 0){
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: `res === undefined` is wrong. If res is really undefined, this will give an error. 

`typeof res === 'undefined'`

File: cdap-ui/app/directives/log-viewer/log-viewer.less [hunk 9]
```diff
@@ -192,6 +238,64 @@ my-log-viewer {
...
+        }
+
+        .application-status.program-completed {
+          .status-circle.program-completed-circle {
+            background-color: @application-completed-color;
+          }
+          color: @application-completed-color;
+        }
+
+        .application-status.program-default {
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.less]: why not nest these inside .application-status?

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 10]
```diff
@@ -14,17 +14,49 @@
  * the License.
  */
 
-function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_ACTIONS, MyCDAPDataSource, $sce, myCdapUrl, $timeout) {
+function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_ACTIONS, MyCDAPDataSource, $sce, myCdapUrl, $timeout, $uibModal, EventPipe) {
   'ngInject';
 
+  window.aa = EventPipe;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: what is this for?

File: cdap-ui/app/directives/log-viewer/log-viewer.less [hunk 11]
```diff
@@ -32,6 +32,48 @@
 @error-color: #f65454;
 @warning-color: #f6b354;
 @selected-row-color: #e4e4e4;
+@application-primary-color: #43b82f;
+@application-secondary-color: #d15668;
+@application-completed-color: #5a84e4;
+@application-default-color: #969696;
+
+body.theme-cdap {
+  .modal.cdap-modal {
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.less:41]: I believe this will node config modal styles. We should target `node-config-modal` with another class so that the namespacing becomes unique for log-viewer. This I believe might affect config modal in hydrator.

**Comment by ajainarayanan** [comment]
@PGuimarais There seems to be build problem. 

```
app/directives/log-viewer/log-viewer.js: line 17, col 146, 'EventPipe' is defined but never used.

1 error
```

File: cdap-ui/app/directives/log-viewer/log-viewer.less (hunk 11 above)
Comment (PGuimarais) [inline cdap-ui/app/directives/log-viewer/log-viewer.less:41]: @ajainarayanan addressed

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 12]
```diff
@@ -492,15 +599,6 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
     return entry;
   };
 
-  $scope.$on('$destroy', function() {
-    if (unsub) {
-      unsub();
-    }
-    if(pollPromise){
-      dataSrc.stopPoll(pollPromise.__pollId__);
-      pollPromise = null;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js:522]: why is this removed? if the program still running and you exit the state, when will the poll get stopped?

also I believe not unsubscribing will cause a memory leak

File: cdap-ui/app/directives/log-viewer/log-viewer.less [hunk 13]
```diff
@@ -32,6 +32,48 @@
...
+          height: ~"-webkit-calc(100vh - 120px)";
+        }
+      }
+    }
+  }
+}
+
+.modal.cdap-modal {
+  &.node-config-modal,
+  &.raw-modal.post-run-actions-modal {
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.less]: why is this modifying post-run-actions-modal and node-config-modal?

File: cdap-ui/app/directives/log-viewer/log-viewer.html [hunk 14]
```diff
@@ -66,7 +75,7 @@
   </my-timeline>
 
   <!-- LOGS TABLE -->
-  <div class="logs-table">
+  <div class="logs-table" ng-class="{'navigator-fullscreen-log-table' : LogViewer.entityName === 'Cloudera Navigator' }">
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.html]: I don't think this should have a specific class at the directive level. Can you target this from the container template?

**Comment by ajainarayanan** [comment]
cdnlogviewer

Two things,
1. The height of the log viewer should be fixed. There are right now two scroll bars and its just makes using logviewer harder.
2. Clicking on the first row (which has a stack trace) pushes the row up with allowing the user to view only the stack trace. And it is also preventing the user from selecting and copy-pasting the stack trace (any click on the stack trace just closes the stack trace)

File: cdap-ui/app/directives/log-viewer/log-viewer.html [hunk 15]
```diff
@@ -35,16 +45,24 @@
...
       <div class="log-action-container">
-        <a  href="{{LogViewer.rawUrl}}"
-            target="_blank" class="log-action-btn view-raw btn btn-default" ng-diabled="!LogViewer.rawUrl.length">
-            View Raw
-        </a>
-        <a class="log-action-btn download-all btn btn-default" ng-click="LogViewer.export()">
-          Download All
-        </a>
+        <button  ng-click="LogViewer.openRaw()"
+                target="_blank"
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.html]: remove target

File: cdap-ui/app/directives/log-viewer/log-viewer.html [hunk 16]
```diff
@@ -19,6 +19,16 @@
   <!-- LOGS HEADERBAR -->
   <div class="logs-headerbar">
     <div class="log-stats-container">
+      <div class="logs-status-container" ng-if="LogViewer.fullScreen">
+        <span class="application-name" ng-bind="LogViewer.entityName">
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.html]: ng-bind and {{}}?

File: cdap-ui/app/directives/log-viewer/log-viewer.html [hunk 17]
```diff
@@ -35,16 +45,24 @@
...
+            <span ng-if="LogViewer.isOpenRaw">
+              <span class="fa fa-spin fa-refresh"></span>
+            </span>
+            <span>View Raw logs</span>
+        </button>
+        <button class="log-action-btn download-all btn btn-default"
+                ng-click="LogViewer.export()"
+                ng-disabled="LogViewer.runId === '' || LogViewer.isDownloading || LogViewer.loading">
+          <span ng-if="LogViewer.isDownloading" ><span class="fa fa-spin fa-refresh"></span></span>
+          <span ng-if="LogViewer.isDownloading">Downloading</span>
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.html]: should combine this into a single ng-if span

File: cdap-ui/app/features/tracker/tracker.less [hunk 18]
```diff
@@ -1143,6 +1143,29 @@ body.state-tracker-integrations {
         right: 9px;
       }
     }
+
+    .log-viewer-parent {
+      position: static;
+      .logs-table {
+        /*
+          60(Integration program heading (MetadataFlow)) +
+          160 (cdn graph) + 82(cdn integration headering ) +
```
Comment (elfenheart) [inline cdap-ui/app/features/tracker/tracker.less]: headering?

**Comment by elfenheart** [comment]
Is the blue start time supposed to work on this branch? It doesn't seem like it does on the cluster

File: cdap-ui/app/directives/log-viewer/log-viewer.html [hunk 19]
```diff
@@ -35,16 +43,23 @@
...
+            <span ng-if="LogViewer.isOpenRaw">
+              <span class="fa fa-spin fa-refresh"></span>
+            </span>
+            <span>View Raw logs</span>
+        </button>
+        <button class="log-action-btn download-all btn btn-default"
+                ng-click="LogViewer.export()"
+                ng-disabled="!LogViewer.displayData.length || LogViewer.isDownloading || LogViewer.loading">
+          <span ng-if="LogViewer.isDownloading" ><span class="fa fa-spin fa-refresh"></span></span>
+          <span ng-if="LogViewer.isDownloading">Downloading</span>
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.html]: consolidate this to be under 1 ng-if?

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 20]
```diff
@@ -39,17 +39,21 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
       case 'SUSPENDED':
+        this.applicationIsRunning = false;
         this.statusType = 1;
         break;
       case 'COMPLETED':
+        this.applicationIsRunning = false;
         this.statusType = 2;
         break;
       default:
+        this.applicationIsRunning = false;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: just use `this.statusType`?

**Comment by elfenheart** [comment]
👍  LGTM
